### PR TITLE
Restrict generational mode to adaptive heuristic

### PR DESCRIPTION
**Pull request by earthling-amzn**
This is a small change to Shenandoah, but it prevents us from running the tests with the generational mode enabled via a jtreg vm option. This requires us to properly integrate the generational mode with the jtreg tests. There are many test files changed, but the `generational` mode is now handled in the same fashion as other modes (`iu`, `passive`). There are two additional advantages to these changes:
* We no longer need to run the test suite twice (once in the default mode and again with `-XX:ShenandoahGCMode=generational`).
* Expanded test coverage, as some tests would spawn their own VMs with permutations of mode and heuristic options that ignored the generational mode.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer) ⚠️ Review applies to [12e402ab](https://git.openjdk.org/shenandoah/pull/233/files/12e402abeb462cc88f4d4b0f6b7933707aa1440b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/233/head:pull/233` \
`$ git checkout pull/233`

Update a local copy of the PR: \
`$ git checkout pull/233` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/233/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 233`

View PR using the GUI difftool: \
`$ git pr show -t 233`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/233.diff">https://git.openjdk.org/shenandoah/pull/233.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/gc/shenandoah/TestParallelRefprocSanity.java [hunk 1]
```diff
@@ -1,50 +0,0 @@
-/*
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/shenandoah/TestParallelRefprocSanity.java:1]: After reference processing was made concurrent, Shenandoah no longer reads the `ParallelRefProcEnabled` flag. Reference processing is covered by `TestRefprocSanity.java`.

File: test/hotspot/jtreg/gc/shenandoah/TestRegionSamplingLogging.java [hunk 2]
```diff
@@ -23,16 +23,26 @@
...
 
 /*
- * @test id=rotation
+ * @test id=default-rotation
  * @requires vm.gc.Shenandoah
  *
  * @run main/othervm -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
  *      -XX:+ShenandoahRegionSampling -XX:+ShenandoahRegionSampling
- *      -Xlog:gc+region=debug:region-snapshots-%p.log::filesize=100,filecount=3
+ *      -Xlog:gc+region=trace:region-snapshots-%p.log::filesize=100,filecount=3
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/shenandoah/TestRegionSamplingLogging.java:31]: This feature of writing the region sampling data out to logs is only on the generational branch at this time. It wouldn't be hard to pull it out into a separate upstream PR.

File: test/hotspot/jtreg/gc/shenandoah/TestSieveObjects.java [hunk 3]
```diff
@@ -90,6 +90,28 @@
...
+ * @library /test/lib
+ *
+ * @run main/othervm -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCHeuristics=adaptive -XX:ShenandoahGCMode=generational
+ *      -XX:+ShenandoahOOMDuringEvacALot -XX:+ShenandoahVerify
+ *      TestSieveObjects
+ *
+ * @run main/othervm -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCHeuristics=adaptive -XX:ShenandoahGCMode=generational
+ *      -XX:+ShenandoahAllocFailureALot -XX:+ShenandoahVerify
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/shenandoah/TestSieveObjects.java:107]: Same rationale for `TestRetainObjects` - are you concerned about the runtime length of the tests?

File: test/hotspot/jtreg/gc/shenandoah/generational/TestCLIModeGenerational.java [hunk 4]
```diff
@@ -23,15 +23,15 @@
...
 
 /*
  * @test TestCLIModeGenerational
  * @requires vm.gc.Shenandoah
  * @summary Test argument processing for -XX:+ShenandoahGCMode=generational.
  * @library /testlibrary /test/lib /
- * @build sun.hotspot.WhiteBox
- * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @build jdk.test.whitebox.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/shenandoah/generational/TestCLIModeGenerational.java:34]: Right - these files were added to the generational branch, so they were not updated when we merged the WhiteBox renaming changes.

File: test/hotspot/jtreg/gc/shenandoah/oom/TestAllocOutOfMemory.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+ * @test id=large
+ * @summary Test allocation of small object to result OOM, but not to crash JVM
+ * @requires vm.gc.Shenandoah
+ * @library /test/lib
+ * @run driver TestAllocOutOfMemory large
+ */
+
+/**
+ * @test id=heap
+ * @summary Test allocation of small object to result OOM, but not to crash JVM
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/shenandoah/oom/TestAllocOutOfMemory.java:35]: Yep, good catch. Thanks.

File: test/hotspot/jtreg/gc/stress/jfr/TestStressBigAllocationGCEventsWithShenandoah.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/*
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply test/hotspot/jtreg/gc/stress/jfr/TestStressBigAllocationGCEventsWithShenandoah.java:1]: I've added one just now.

File: test/hotspot/jtreg/gc/shenandoah/generational/TestCLIModeGenerational.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/generational/TestCLIModeGenerational.java:34]: Okay, great. Nevermind then.

File: test/hotspot/jtreg/gc/shenandoah/TestAllocObjectArrays.java [hunk 7]
```diff
@@ -134,6 +161,11 @@
  *      -XX:+UseShenandoahGC
  *      -XX:-UseTLAB -XX:+ShenandoahVerify
  *      TestAllocObjectArrays
+ *
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -Xmx1g -Xms1g
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCHeuristics=adaptive -XX:ShenandoahGCMode=generational
+ *      -XX:-UseTLAB -XX:+ShenandoahVerify
```
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/TestAllocObjectArrays.java:167]: Ah right, it differs from the above by mode. Nevermind!

File: test/hotspot/jtreg/gc/shenandoah/TestRegionSamplingLogging.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/TestRegionSamplingLogging.java:31]: `TestRegionSamplingLogging` is only in generational branch? Then disregard this comment :)

File: test/hotspot/jtreg/gc/shenandoah/TestSieveObjects.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/TestSieveObjects.java:107]: Yes, same overlook on my side. This is fine.

File: test/hotspot/jtreg/gc/shenandoah/compiler/BarrierInInfiniteLoop.java [hunk 8]
```diff
@@ -22,19 +22,40 @@
...
  * @requires vm.gc.Shenandoah
  *
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+UseShenandoahGC -Xcomp -XX:CompileOnly=BarrierInInfiniteLoop::test1
  *                   -XX:CompileOnly=BarrierInInfiniteLoop::test2 -XX:CompileOnly=BarrierInInfiniteLoop::test3 -XX:CompileCommand=quiet BarrierInInfiniteLoop
- * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+UseShenandoahGC -XX:ShenandoahGCMode=iu -Xcomp -XX:CompileOnly=BarrierInInfiniteLoop::test1
- *                   -XX:CompileOnly=BarrierInInfiniteLoop::test2 -XX:CompileOnly=BarrierInInfiniteLoop::test3 -XX:CompileCommand=quiet BarrierInInfiniteLoop
- *
  */
 
+ /**
```
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/compiler/BarrierInInfiniteLoop.java:35]: Yes, but this test chases the particular compiler bug. Running in multiple modes with `-Xcomp` and `-XX:CompileOnly` might not be fast, so it is better to keep these regression tests targeted. It is not the issue for compiler tests that are not regression tests, like `TestCloneBarrier`.

[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/jtreg/gc/shenandoah/compiler/BarrierInInfiniteLoop.java:35]: I still don't think it is a good idea to touch these regression tests.

File: test/hotspot/jtreg/gc/shenandoah/compiler/TestReferenceCAS.java [hunk 9]
```diff
@@ -53,32 +53,6 @@
  * @run main/othervm -Diters=20000 -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -XX:ShenandoahGCHeuristics=aggressive -XX:+UseShenandoahGC -XX:-UseCompressedOops -XX:TieredStopAtLevel=4  TestReferenceCAS
  */
 
-/*
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/shenandoah/compiler/TestReferenceCAS.java]: This one is actually not a regression test, it is similar to `TestClone` that verifies the generic support for Shenandoah barriers. we should keep it here.

**Review by shipilev — APPROVED** [review]
Everything is good now, except for `TestReferenceCAS`.

**Pull request closed without merging**